// moo-component-overlay/Gruntfile.js
var webpack = require('webpack');

module.exports = function(grunt) {

  require('load-grunt-tasks')(grunt);

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    watch: {
      css: {
        files: ['lib/*.css', 'test/src/*.css'],
        tasks: ['css']
      },
      js: {
        files: ['lib/*.js'],
        tasks: ['js']
      }
    },
    postcss: {
      options: {
        bemLinter: true,
        compress: true,
        browsers: [
          'Last 2 versions',
          'IE >= 9'
        ]
      },
      compile: {
        files: {
          'test/style.css': ['./test/src/index.css']
        }
      }
    },
    webpack: {
      compile: {
        entry: "./test/src/index.js",
        output: {
          path: "test",
          filename: "scripts.js",
        },
        plugins: [
          new webpack.ProvidePlugin({
            $: 'jquery',
            jQuery: 'jquery',
            'window.jQuery': 'jquery'
          }),
          new webpack.optimize.DedupePlugin(),
          new webpack.optimize.UglifyJsPlugin()
        ]
      }
    }
  });

  grunt.registerTask('default', ['compile']);
  grunt.registerTask('css', ['postcss:compile']);
  grunt.registerTask('js', ['webpack:compile']);
  grunt.registerTask('compile', ['css', 'js']);
};
