// moo-component-header/lib/nav-dropdown.js
var $ = require('jquery');
var Openable = require('moo-jsutil-openable');

var NavDropdown = (function () {
    
    $.support.transition = (function(){
        var thisBody = document.body || document.documentElement,
            thisStyle = thisBody.style,
            support = thisStyle.transition !== undefined || thisStyle.WebkitTransition !== undefined || thisStyle.MozTransition !== undefined || thisStyle.MsTransition !== undefined || thisStyle.OTransition !== undefined;
        return support;
    })();
    var cssTransitionsDetection = $.support.transition;
    
  var navDropdown = {};

    navDropdown = {
        settings: {
          section:            undefined,
          clicked:            false,
          clickedTitle:       undefined,
          clickedSection:     undefined,
          clickedSectionRow:  undefined,
          transitions:        cssTransitionsDetection
        },
        animationName: {
           fadeInDownMenu:  "fadeInDownMenu",
           slideDownMenu: "slideDownMenu",
           fadeIn: "fadeIn",
           fadeOutUpMenu: "fadeOutUpMenu",
           slideUpMenu: "slideUpMenu"
        },
        init: function() {
            this.bindActions();
            this.resizeEvents();
        },
    
        bindActions: function() {
          $(document).on('click', '.Navigation, .overlay-scroll', this.keepMenu);
          $(document).on('click', '.has-dropdown', this.toggleMenu);
          $(document).on('click', 'html', this.hideMenu);
          $(document).on('click', '.Header-toggleLink', this.toggleMobileMenu);
        },
    
        keepMenu: function(e){
          e.stopPropagation();
        },
        
        hideMenu: function() {
          var s = navDropdown.settings;
          var a = navDropdown.animationName;

          s.clickedTitle      = '.Menu-title--' + s.section;
          s.clickedSection    = '.Menu-section--' + s.section;
          s.clicked           = false;
        
           if($(s.clickedTitle).parent().hasClass('is-open')) {
               $('.MainNav-item').removeClass('is-open');
               
               if(s.transitions === true) {
                    $('.Navigation-dropdownContainer').addClass('fadeOutUpMenu').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                        if (e.originalEvent.animationName === a.fadeOutUpMenu) {
                            $(this).removeClass('fadeOutUpMenu is-open');
                            $(s.clickedSection).removeClass('is-open');
                            $('.Header').removeClass('is-open')
                            $('.Navigation-dropdownContainer .Container').empty();
                        }
                    });
              } else {
                    $('.Navigation-dropdownContainer').removeClass('is-open');
                    $('.Navigation-dropdownContainer .Container').empty();
              }
          }
        },
    
        toggleMenu: function(e) {
            e.preventDefault();
            var s = navDropdown.settings;
            var a = navDropdown.animationName;
            
            s.section           = $(this).attr('data-menu-dropdown');
            s.clickedTitle      = '.Menu-title--' + s.section;
            s.clickedSection    = '.Menu-section--' + s.section;
            windowWidth         = $(window).width();
            
            // Open menu
            if($(s.clickedTitle).parent().hasClass('is-open') === false) {
              
              // Clone sub menu and add to dropdown container
              if ($('.Navigation-dropdownContainer ' + s.clickedSection).length === 0) {    
                  $('.MainNav ' + s.clickedSection).clone().appendTo('.Navigation-dropdownContainer .Container');
              }
            
              // Close any currently is-open
              $('.MainNav-item').removeClass('is-open');
              $('.MainNav-itemDropdown').removeClass('is-open');
              $(s.clickedSection).removeClass('is-open');
            
              // is-open the clicked version
              $(s.clickedTitle).parent().addClass('is-open');
              $(s.clickedSection).addClass('is-open');
            
              // is-open and Animate
              $('.Navigation-dropdownContainer').addClass('is-open');
              
              // Display menu
              if (s.clicked === false) {
                  if (s.transitions === true) {
                        // Desktop Menu
                        if (windowWidth >= 981) {
                            $('.Navigation-dropdownContainer').addClass('fadeInDownMenu').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                                if (e.originalEvent.animationName === a.fadeInDownMenu) {
                                    $(this).removeClass('fadeInDownMenu');
                                    $('body, .Header').addClass('is-open');
                                }
                            });
                        }
                        // Mobile Menu
                        else {
                            $('.MainNav ' + s.clickedSection).addClass('slideDownMenu').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                                if (e.originalEvent.animationName === a.slideDownMenu) {
                                    $(this).removeClass('slideDownMenu');
                                    $('.Navigation-dropdownContainer').removeClass('fadeInDownMenu');
                                }
                            });   
                        }                     
                  }
              // Toggle between menus
              } else {
                  if (s.transitions === true) {
                        // Desktop Menu
                        if (windowWidth >= 981) {
                            $('.Navigation-dropdownContainer ' + s.clickedSection).addClass('fadeIn').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                                if (e.originalEvent.animationName === a.fadeIn) {
                                    $(this).removeClass('fadeIn');
                                    $('body, .Header').addClass('is-open');
                                }
                            });
                        }
                        // Mobile Menu
                        else {
                            $('.MainNav ' + s.clickedSection).addClass('fadeInFast').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                                if (e.originalEvent.animationName === a.fadeIn) {
                                    $(this).removeClass('fadeInFast');
                                    $('.Navigation-dropdownContainer ' + s.clickedSection).removeClass('fadeIn');
                                }
                            });
                        }
                  }
              }
            
              s.clicked = true;
            
            // Close menu
            } else {
              $('.MainNav-item').removeClass('is-open');
              
              if (s.transitions === true) {
                    // Desktop Menu
                    if (windowWidth >= 981) {
                        $('.Navigation-dropdownContainer').addClass('fadeOutUpMenu').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                            if (e.originalEvent.animationName === a.fadeOutUpMenu) {
                                $(this).removeClass('fadeOutUpMenu is-open');
                                $('.Navigation-dropdownContainer ' + s.clickedSection).removeClass('is-open');
                                $('body, .Header').removeClass('is-open');
                                s.clicked = false;
                            }
                        });
                    }
                    // Mobile Menu
                    else {
                        $('.MainNav ' + s.clickedSection).addClass('slideUpMenu').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                            if (e.originalEvent.animationName === a.slideUpMenu) {
                                $(this).removeClass('slideUpMenu is-open');
                                $('.Navigation-dropdownContainer').removeClass('fadeOutUpMenu is-open');
                                $('.Navigation-dropdownContainer ' + s.clickedSection).removeClass('fadeIn is-open');
                            }
                        });   
                    }                  
              } else {
                  $('.Navigation-dropdownContainer').removeClass('is-open');
                  $(s.clickedSection).removeClass('is-open');
                  
                  s.clicked = false;
              }
            }
        },
          
        toggleMobileMenu: function(e) {
          e.preventDefault();
          var windowWidth = $(window).width();
          var a = navDropdown.animationName;
          
          // If menu is not open
          if (!$('.Header-navigation').hasClass('is-open')) {
              
              // Mobile
              if (windowWidth <= 767) {
        	      // Fade out the wordMark
        	      $('.Header-wordMark').addClass('fadeOutFast').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
            	       if (e.originalEvent.animationName === a.fadeOut) {
                	       $(this).removeClass('fadeOutFast').hide();
                	   }
                  });
              }
              // Fade in the icon toggle
              $('.Header-iconToggle').addClass('fadeInFast').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
        	      if (e.originalEvent.animationName === a.fadeIn) {
                     $(this).removeClass('fadeInFast');
                  }
              });
              // Slide down and show the menu
              $('.Header-navigation').addClass('is-open slideDownMenu').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
        	      if (e.originalEvent.animationName === a.slideDownMenu) {
                      $(this).removeClass('slideDownMenu');
        	      }
              });
          // If menu is open
          } else {
              // Mobile
              if (windowWidth <= 767) {
        	      // Fade in the hidden wodMark
        	      $('.Header-wordMark').show().addClass('fadeInFast').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
            	       if (e.originalEvent.animationName === a.fadeIn) {
                	       $(this).removeClass('fadeInFast');
                	   }
                  });
              // Tablet and above
              } else {
                  // If wordMark is hidden show it
                  // since we hide at the mobile breakpoint
                  if ($('.Header-wordMark').css('display') == 'none') {
                      $('.Header-wordMark').show().addClass('fadeInFast').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                	       if (e.originalEvent.animationName === a.fadeIn) {
                    	       $(this).removeClass('fadeInFast');
                    	   }
                      });
                  }
              }
              // Fade out icon toggle
              $('.Header-iconToggle').addClass('fadeOutFast').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
        	      if (e.originalEvent.animationName === a.fadeOut) {
                    $(this).removeClass('fadeOutFast');
                  }
              });
              // Slide up and hide the menu
              $('.Header-navigation').addClass('slideUpMenu').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
        	      if (e.originalEvent.animationName === a.slideUpMenu) {
                      $(this).removeClass('slideUpMenu is-open');
        	      }
              });
          }
        },
        resizeEvents: function() {
            var openMobile = false;
            var a = navDropdown.animationName;
            
            // Set variable to track if the menu
            // was opened at the mobile breakpoint
            // using the Header-iconToggle element
            $('.Header-iconToggle').click(function() {
                openMobile = !openMobile; 
            });
              
            $(window).on('resize', function(){
                var windowWidth = $(window).width();
                
                // If menu is resized to tablet
                // from mobile and the wordMark
                // and is hidden, show it
                if (windowWidth < 981 && windowWidth > 766) {
                    if ($('.Header').hasClass('is-open') && $('.Header-wordMark').css('display') == 'none') {
                          $('.Header-wordMark').show().addClass('fadeInFast').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                    	       if (e.originalEvent.animationName === a.fadeIn) {
                        	       $(this).removeClass('fadeInFast');
                        	   }
                          });
                      }
                } 
                // If menu is resized to mobile
                // from tablet and the wordMark
                // and is visible, hide it
                else if (windowWidth < 767) {
                    if ($('.Header').hasClass('is-open') && $('.Header-wordMark').css('display') == 'inline-block') {
                      $('.Header-wordMark').addClass('fadeOutFast').one('webkitAnimationEnd mozAnimationEnd MSAnimationEnd oanimationend animationend', function(e){
                	       if (e.originalEvent.animationName === a.fadeOut) {
                    	       $(this).removeClass('fadeOutFast').hide();
                    	   }
                      });
                    }
                } 
                
                // If menu was opened at mobile
                // then resized to desktop
                // and then resized to mobile
                // make sure menu is still open    
                if (windowWidth < 767 && openMobile) {
                    $('.Header').addClass('is-open');    
                }
            });
        }
    };

  return navDropdown;

})();

module.exports = NavDropdown;
